contracts/programs/keystone-forwarder/src/lib.rs

```diff
@@ -0,0 +1,504 @@
+    pub proposed_owner: Signer<'info>,
+}
+
+// combination of solidity's OracleSet and the configId mapping
+#[account]
+#[derive(Default)]
+pub struct OraclesConfig {
+    config_id: u64,
+    f: u8,
+    signer_addresses: Vec<EthAddress>, // maybe think about doing ECDSA over ed25519?
+} 
+
+impl OraclesConfig {
+    pub const INIT_SPACE: usize = 8 + 1 + 4;
+
+    pub fn space_with_signers(num_signers: usize) -> usize {
+        Self::INIT_SPACE + num_signers*20
+    }
+}
+
+fn get_config_id(don_id: u32, config_version: u32) -> u64 {
+    ((don_id as u64) << 32) | (config_version as u64)
+}
+
+#[derive(Accounts)]
+#[instruction(don_id: u32, config_version: u32, signer_addresses: Vec<EthAddress>)]
+pub struct InitOraclesConfig<'info> {
+
+    pub state: Account<'info, ForwarderState>,
+
+    #[account(
+        init,
+        payer = owner,
+        seeds = [b"config", state.key().as_ref(), &get_config_id(don_id, config_version).to_be_bytes()],
+        bump,
+        space = OraclesConfig::space_with_signers(signer_addresses.len())
+    )]
+    oracles_config: Account<'info, OraclesConfig>,
+
+    #[account(mut, address = state.owner @ AuthError::Unauthorized)]
+    pub owner: Signer<'info>,  // must be the same owner as the one in the state account
+
+    pub system_program: Program<'info, System>,
+}
+
+#[derive(Accounts)]
+#[instruction(don_id: u32, config_version: u32, signer_addresses: Vec<Pubkey>)]
+pub struct UpdateOraclesConfig<'info> {
+
+    pub state: Account<'info, ForwarderState>,
+
+    #[account(
+        mut,
+        seeds = [b"config", state.key().as_ref(), &get_config_id(don_id, config_version).to_be_bytes()],
+        bump,
+        realloc = ANCHOR_DISCRIMINATOR + OraclesConfig::space_with_signers(signer_addresses.len()),
+        realloc::payer = owner,
+        realloc::zero = true
+    )]
+    oracles_config: Account<'info, OraclesConfig>,
+
+    #[account(mut, address = state.owner @ AuthError::Unauthorized)]
+    pub owner: Signer<'info>,
+
+    pub system_program: Program<'info, System>,
+}
+
+#[derive(Accounts)]
+#[instruction(don_id: u32, config_version: u32)]
+pub struct CloseOraclesConfig<'info> {
+
+    pub state: Account<'info, ForwarderState>,
+
+    #[account(
+        mut,
+        seeds = [b"config", state.key().as_ref(), &get_config_id(don_id, config_version).to_be_bytes()],
+        bump,
+        close = owner
+    )]
+    oracles_config: Account<'info, OraclesConfig>,
+
+    #[account(mut, address = state.owner @ AuthError::Unauthorized)]
+    pub owner: Signer<'info>,  // must be the same owner as the one in the state account
+}
+
+// data =  bump | len_signatures (N) | signatures (N*65) | raw_report (M) | report_context (96)
+fn extract_raw_report(data: &[u8]) -> Vec<u8> {
+    let _execution_state_bump = data[0];
+    let len = data[1] as usize;
+    let data = &data[2..];
+    let _signatures = &data[..65*len];
+    let data = &data[65*len..];
+    let _report_context = &data[data.len()-96..];
```

Comment:
I think it would make more sense to have the report context before the report

Reply:
I kept it this way because the existing solana keyring assumes the report context comes after https://github.com/smartcontractkit/chainlink/blob/develop/core/services/keystore/keys/ocr2key/solana_keyring.go#L38 . But open to changing it as well. 

We should be using ocr2 signatures, not ocr3 right?